content/blog/debugging-valgrind-gdb.md

```diff
@@ -0,0 +1,190 @@
+---
+title: "Debugging memory errors with valgrind and gdb"
+author: Rich FitzJohn
+date: 2021-06-10
+tags:
+ - dust
+ - R
+ - debugging
+---
+
+We've already written about debugging R packages with valgrind and gdb together [here](https://reside-ic.github.io/blog/debugging-and-fixing-crans-additional-checks-errors/). This post shows how to use them together so that you can use gdb to inspect the state of a program at the point where valgrind has noticed that there's an error.  The process is a bit weird, but does work!  This will likely only work on Linux (not windows or macOS).
+
+## The problem
+
+We noticed that as of 0.9.3 of `dust` we had a `std::bad_alloc` error when running models with the `float` type (rather than `double`). After a bit of narrowing it down we managed to reduce the problem to this bit of code:
+
+```
+sirs <- dust::dust_example("sirs")
+pars <- list(beta = 0.2, gamma = 0.1, alpha = 0.1, freq = 4)
+end <- 150 * 4
+steps <- seq(0, end, by = 4)
+ans <- sirs$new(pars, 0, 1, seed = 1L)$simulate(steps)
+dat <- data.frame(step = steps[-1],
+                  date = steps[-1] / 4,
+                  incidence = ans[4, 1, -1])
+dat_dust <- dust::dust_data(dat, "step")
+
+path_sirs <- system.file("examples/sirs.cpp", package = "dust")
+sirs_gpu <- dust::dust(path_sirs, gpu = FALSE, real_t = "float")
+
+set.seed(1)
+
+mod <- sirs_gpu$new(list(), 0, 100, seed = 42L)
+mod$set_data(dat_dust)
+for (i in 1:20) {
+  message(i)
+  mod$reset(list(), 0)
+  mod$filter()
+}
+```
+
+This will crash deterministic ally on the 18th iteration through the loop.  Once we narrowed this down, we could see from valgrind where the error was coming from (running the script above with `R -d valgrind -f script.R`):
```

Comment:
```suggestion
This will crash deterministically on the 18th iteration through the loop.  Once we could reliable reproduce the error, we could see from valgrind where the error was coming from (running the script above with `R -d valgrind -f script.R`):
```

Reply:
Don't know if it's worth saying something about the advantage of dust design, that the RNG seeding really helps avoid stochastic errors